test/types/queries.test.ts

```diff
@@ -307,20 +313,34 @@ function autoTypedQuery() {
 }
 
 function gh11964() {
+  class Repository<T extends { id: string }> {
+    find(id: string) {
+      const idCondition: Condition<T['id']> = id as Condition<T['id']>;
+
+      // `as` is necessary because `T` can be `{ id: never }`,
+      // so we need to explicitly coerce
+      const filter: FilterQuery<T> = { id } as FilterQuery<T>;
```

Comment:
can `T` really be `{ id: never }` if it is now explicitly set as `T extends { id: string }`? is this some kind of typescript magic i am missing?

Reply:
@hasezoey @vkarpov15 I have just tested it and checked that even writing `const idCondition: Condition<T['id']> = id` without the `as` works fine.

The 2nd one fails without the `as` keyword because [in typescript, a concrete instance is not allowed to be assigned to a type parameter](https://stackoverflow.com/questions/56505560/how-to-fix-ts2322-could-be-instantiated-with-a-different-subtype-of-constraint#:~:text=In%20typescript%2C%20a%20concrete%20instance%20is%20not%20allowed%20to%20be%20assigned%20to%20a%20type%20parameter.). This means that even if you try to write `const obj: T = { id: string };`, then it will fail. Unless you add the `as` keyword and change it to `const obj: T = { id: string } as T;` Thus, we also need to add the `as` keyword for the 2nd statement, so that it follows typescript rule. Weird behaviour by typescript, but what's not weird about typescript :)

Reply:
@hasezoey yep try this TypeScript playground: https://www.typescriptlang.org/play?#code/PTAEGUAsHsHcGdQBdIEMkBpQFMBu2A7UASwDNQBPaAV1FQBN7QADAMlAG8T6AuUeJACdiBAOagAvsywpsoQdnjUANkhHikFAA5z60RQQDkSUAWwBjRfFTDlFAFAhQafHSLNivfkPXNQWwWgdQU17TR1QABVQAF5Obj4zfEFJUHYuTz4BYTFJAG57c2gCAVAADz5ouIyvQ1JoaEN8+0cwKDhEFHQWTNM8bEE-AWJlZRwypEJ6RA8vbN9C5VR4RAAlbC1oeGIkaEEKAB5o7AmpxBqsn1yJAD5Oe1BQUhF6AApe+bEASnvHx6KSiZSABGSqxeKeZqPCT2GEA0oUPjrTbbXb7A4cXpJAa3cFmWCgZFbHZ7Q6YrzYwS3V5fPJAA . It is very weird that `{ id: never }` does seem to satisfy `extends { id: string }` but that's TypeScript for you :)

@FaizBShah thanks for the explanation :+1:

Reply:
Yupp!! Glad I could help :D

Reply:
@vkarpov15 your example seemed to have errors, but i could reproduce it with [just this](https://www.typescriptlang.org/play?#code/PTAEGUAsHsHcGdQBdIEMmgAYEsAmAuUAOwFMA3EgJ01HiWwBsHQSAPJEo3RHA2pStiIBzTACgAxg1TxEAJRIAHaPGxJolAJ4AeACot2nbqADeoPITqCRoAL4A+U2NCgAZkNwAKC-2vCAlE4uLs52YraS0ER0oJqECsqq6lraJj6kFJQOoAC8xCSwoAkqaho6aXwZVA6e-gDcQA).